dotCMS/src/main/java/com/dotmarketing/cms/urlmap/URLMapAPIImpl.java

```diff
@@ -0,0 +1,318 @@
+import com.dotmarketing.util.UtilMethods;
+import com.liferay.util.StringPool;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.TreeSet;
+import org.apache.lucene.queryparser.classic.QueryParser;
+import org.jetbrains.annotations.NotNull;
+
+public class URLMapAPIImpl {
+
+    private Collection<PatternCache> patternsCache;
+
+    private final UserWebAPI wuserAPI = WebAPILocator.getUserWebAPI();
+    private final HostWebAPI whostAPI = WebAPILocator.getHostWebAPI();
+    private final ContentletAPI contentletAPI = APILocator.getContentletAPI();;
+    private final PermissionAPI permissionAPI = APILocator.getPermissionAPI();
+    private final IdentifierAPI identifierAPI = APILocator.getIdentifierAPI();
+
+    public URLMapInfo processURLMap(final UrlMapContext context)
+            throws DotSecurityException, DotDataException {
+
+        if (this.shouldLoadPatterns()) {
+            this.loadPatterns();
+        }
+
+        if (containsRegEx(context.getUri())) {
+            final Matches matches = this.findPatternChange(context.getUri());
+
+            final Structure structure = CacheLocator.getContentTypeCache()
+                    .getStructureByInode(matches.getPatternChange().getStructureInode());
+
+            final Field hostField = this.findHostField(structure);
+
+            final Contentlet contentlet = this.getContentlet(matches, structure, hostField, context);
+
+            if (contentlet == null) {
+                return null;
+            }
+
+            final Identifier pageUriIdentifier = this.getDetailtPageUri(structure);
+
+            return new URLMapInfo(contentlet, pageUriIdentifier);
+        } else {
+            return null;
+        }
+    }
+
+    private Identifier getDetailtPageUri(final Structure structure) {
+        if (structure != null && UtilMethods.isSet(structure.getDetailPage())) {
+
+            try {
+
+                final Identifier identifier = this.identifierAPI.find(structure.getDetailPage());
+                if (identifier == null || !UtilMethods.isSet(identifier.getInode())) {
+                    throw new DotRuntimeException(
+                            "No valid detail page for structure '" + structure.getName()
+                                    + "'. Looking for detail page id=" + structure
+                                    .getDetailPage());
+                }
+
+                return identifier;
+            } catch (Exception e) {
+                throw new DotRuntimeException(e);
+            }
+        } else {
+            return null;
+        }
+    }
+
+    private boolean containsRegEx(final String uri) {
+        try {
+            final String mastRegEx = CacheLocator.getContentTypeCache().getURLMasterPattern();
+
+            final String url = !uri.endsWith(StringPool.FORWARD_SLASH) ? uri + StringPool.FORWARD_SLASH : uri;
+            return RegEX.contains(url, mastRegEx);
```

Comment:
not sure if this should be case sensitive or not, however take in consideration I have added a method to do insensitive contains I think so

Reply:
Really it is not necessary in this case